core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,428 @@
+   *           attribute values for that expression
+   */
+  def executeQueryPlan(workspaceId: UUID,
+                       entityType: String,
+                       entityName: String,
+                       rootEntityType: String,
+                       plan: QueryPlan,
+                       entityLookups: Seq[ExpressionLookup] = Seq.empty
+  )(implicit
+    executionContext: ExecutionContext
+  ): Future[Seq[ExpressionAndResult]] =
+    repository.dataSource
+      .inTransaction { _ =>
+        if (plan.relationChain.isEmpty && entityLookups.isEmpty) {
+          repository.queries.getEntity(workspaceId, entityType, entityName).map {
+            case Some(entity) => Map(entity.name -> entity)
+            case None =>
+              throw new RawlsExceptionWithErrorReport(
+                ErrorReport(StatusCodes.NotFound, s"Entity of type $entityType with name $entityName not found.")
+              )
+          }
+        } else {
+          repository.queries.queryRelatedRecordsWithArray(workspaceId, entityType, entityName, plan.relationChain)
+        }
+      }
+      .map { entityRecords =>
+        // Validate entity types if we have entityLookups
+        if (entityLookups.nonEmpty && entityRecords.nonEmpty) {
+          val actualEntityTypes = entityRecords.values.map(_.entityType).toSet
+          if (actualEntityTypes.nonEmpty && !actualEntityTypes.contains(rootEntityType)) {
+            val actualTypesStr = actualEntityTypes.mkString(", ")
+            throw new RawlsExceptionWithErrorReport(
+              ErrorReport(
+                StatusCodes.BadRequest,
+                s"The expression in your SubmissionRequest matched only entities of the wrong type. " +
+                  s"(Expected type $rootEntityType, but got $actualTypesStr.)"
+              )
+            )
+          }
+        }
+
+        // For each expression in this query plan, create an ExpressionAndResult
+        plan.expressionMappings.toSeq.flatMap { case (expression, attributeNames) =>
+          attributeNames.map { attrName =>
+            val attributeName = AttributeName.fromDelimitedName(attrName)
+
+            if (entityType == rootEntityType) {
+              // Entity type is the root entity type: aggregate all attributes and map to the original entity name
+              val attrs: Seq[AttributeValue] =
+                entityRecords.values.toSeq.flatMap(_.toEntity.attributes.get(attributeName)).flatMap {
+                  case avl: AttributeValueList => avl.list
+                  case av: AttributeValue      => Seq(av)
+                  case _                       => Seq.empty
+                }
+              (expression, Map(entityName -> Success(attrs)))
+            } else {
+              // Entity type is not root entity type, e.g. we're dealing with a set: map to actual entity names from the query results
+              val entityToAttributeValues = entityRecords.map { case (actualEntityName, record) =>
```

Comment:
Note that the key in `entityRecords` should be the name of the root entity, not the name of the entity which is the value in the map. And the value should go back to being a Seq